src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1462,6 +1458,29 @@ AdjustReadIntermediateResultCost(RangeTblEntry *rangeTableEntry, RelOptInfo *rel
 }
 
 
+/*
+ * CopyOuterPlanParamLists deep-copies plan_params lists of the outer nodes
+ * of the given root, and returns them in a list of lists of PlannerParamItem.
+ * First item in this list corresponds to the parent node, and the last
+ * item corresponds to the outer most node.
+ */
+static List *
+CopyOuterPlanParamLists(PlannerInfo *root)
```

Comment:
Naming PlannerInfo variables as `root` to match PostgreSQL's convention. See the comments for `PlannerInfo` in `relation.h`.

Reply:
Also note that keeping a copy of plan_params lists seems necessary, for some reason that I'm not sure of, all `PlannerInfo->planner_params` lists are empty in `relation_restriction_equivalence.c`.

Reply:
> Also note that keeping a copy of plan_params lists seems necessary

Just to clarify: Before the patch, we already keep the copies that are only one level up, right? So, we're only keeping more stuff when necessary?

> , for some reason that I'm not sure of, all PlannerInfo->planner_params lists are empty in relation_restriction_equivalence.c

I think that's because we're passing `baseRelOptInfo->subroot` specific per subquery, not the `root` that we see in the relation restriction hook. It looks like we've somewhat related discussion sometime [ago](https://github.com/citusdata/citus/pull/1268/files#r108345915).

Reply:
> > Also note that keeping a copy of plan_params lists seems necessary
> 
> Just to clarify: Before the patch, we already keep the copies that are only one level up, right? So, we're only keeping more stuff when necessary?

Yes, before this patch we were keeping one level up, now we keep all the levels. I'm not sure if we can decide at this point if a levels are necessary or not, so just keeping all levels up.

I'm looking to see if we can avoid keeping all levels and only the necessary ones ...

Reply:
Today I looked into if it is possible to not copy any `plan_params` lists and make the code simpler. I haven't found a solution which works on all cases yet.

Using `PlannerInfo->simple_rel_array[idx]->subplan_params` was close, but it was failing one of the tests.

If this change isn't urgent, maybe we can keep this on hold for few more days so I can come back and look into it again.

Reply:
> Today I looked into if it is possible to not copy any plan_params lists and make the code simpler.
> If this change isn't urgent, maybe we can keep this on hold for few more days so I can come back and look into it again.

Well, I don't mind copying those params :) But, still getting this change to 8.3 should be sufficient (as far as I can see), so feel free to look later

Reply:
> I'm looking to see if we can avoid keeping all levels and only the necessary ones ...

We could maybe only copy the params when `IsA(originalParamItem->item, Var)` because we only need them. But, I don't think it is worth doing that at the moment, you could maybe consider adding a comment if you want.

But, we typically have very little params, so I'm fine with what we have here. My main motivation to avoid optimizing here is that as the number of params increase, the execution times typically increases a lot. So, saving very little here isn't very tempting to me.

Reply:
> I'm looking to see if we can avoid keeping all levels and only the necessary ones ...
> 
> We could maybe only copy the params when IsA(originalParamItem->item, Var) because we only need them. But, I don't think it is worth doing that at the moment, you could maybe consider adding a comment if you want.
> But, we typically have very little params, so I'm fine with what we have here. My main motivation to avoid optimizing here is that as the number of params increase, the execution times typically increases a lot. So, saving very little here isn't very tempting to me.

Added a comment.